New subcommand `spack config scopes [-hips]`

greenc-FNAL opened the pull request:
List the scopes available in the current environment in descending priority order.

* With `-p|--path-scopes`, list only writable, file-based scopes (ignoring `command_line` and `_builtin`).
* With `-i|--included`, list only included scopes (specified in `include.yaml` or an environment `include:` clause).
* With `-s|--show-paths`, annotate scopes with their directory or file (with non-option argument `section`) locations when appropriate.
* Automatically takes account of `SPACK_DISABLE_LOCAL_CONFIG=true`

Useful for understanding the best scope in which to put a particular specific configuration item, or for debugging "missing" config.

### Examples

* MacOS:
  ```console
  $ spack arch
  darwin-sequoia-m1
  $ spack config scopes
  command_line user site system defaults defaults:darwin defaults:base _builtin
  ```

* Linux:
  ```console
  $ spack arch
  linux-almalinux9-cascadelake
  $ spack config scopes
  command_line user site system defaults defaults:linux defaults:base _builtin
  $ SPACK_DISABLE_LOCAL_CONFIG=1 spack config scopes
  command_line site defaults defaults:linux defaults:base _builtin
  $ SPACK_DISABLE_LOCAL_CONFIG=1 spack config --scope=site scopes --path-scopes
  site defaults defaults:linux defaults:base
  $ SPACK_DISABLE_LOCAL_CONFIG=1 spack config --scope=site scopes --included
  defaults:linux defaults:base
  $ SPACK_DISABLE_LOCAL_CONFIG=1 spack config --scope=site scopes --included --show-paths
  defaults:linux (/home/greenc/work/cet-is/sources/Spack/spack/run/etc/spack/defaults/linux/) defaults:base (/home/greenc/work/cet-is/sources/Spack/spack/run/etc/spack/defaults/base/)
  $ SPACK_DISABLE_LOCAL_CONFIG=1 spack config --scope=site scopes modules
  command_line site (/home/greenc/work/cet-is/sources/Spack/spack/run/etc/spack/) defaults (/home/greenc/work/cet-is/sources/Spack/spack/run/etc/spack/defaults/modules.yaml) defaults:linux (/home/greenc/work/cetis/sources/Spack/spack/run/etc/spack/defaults/linux/modules.yaml) defaults:base (/home/greenc/work/cet-is/sources/Spack/spack/run/etc/spack/defaults/base/) builtin
  ```

greenc-FNAL committed:
New command list-scopes [--file] [--non-platform]

Add missing `_non_platform_scopes()`

Update shell completion scripts

Scope is "defaults," not, "default."

Reorder tests to look for OS-dependent test hysteresis

Disable output suppression for pipeline failure diagnosis

Remove problematic test criteria

`defaults` apparently not available in test environment

greenc-FNAL commented:
@tldahlgren I have (I believe) brought this PR up to date in the light of #48784. I'd be grateful if you could help me get this merged, please.

tldahlgren commented on lib/spack/spack/cmd/config.py:
```diff
@@ -68,6 +68,18 @@ def setup_parser(subparser):
     sp.add_parser("list", help="list configuration sections")
 
+    list_scopes_parser = sp.add_parser("list-scopes", help="list defined scopes")
+    stype = list_scopes_parser.add_mutually_exclusive_group(required=False)
```
I know there aren't a lot of scopes by default, but why are these mutually exclusive?

greenc-FNAL replied:
I think because the way I was originally obtaining non-platform scopes, they were `--file` by construction.

tldahlgren reviewed (commented):
LGTM. I have a couple of questions. One is about the mutual exclusion of the options. The other is have you confirmed the outputs in the description are up-to-date?

tldahlgren commented on lib/spack/spack/test/cmd/config.py:
```diff
@@ -38,6 +38,24 @@ def config_yaml_v015(mutable_config):
+    assert "_builtin" not in output
+
+
+def test_config_list_scopes_non_platform():
+    output = config("list-scopes", "--non-platform")
+    assert "site" in output
+    assert "user" in output
```
This test fails for me if `SPACK_DISABLE_LOCAL_CONFIG` is `true` when I run the tests.

tldahlgren replied:
You might be able to get some inspiration from https://github.com/spack/spack/blob/8ac826cca892f9060ced53a071e78e0119dd8aca/lib/spack/spack/test/config.py#L1230.

greenc-FNAL commented:
> LGTM. I have a couple of questions. One is about the mutual exclusion of the options. The other is have you confirmed the outputs in the description are up-to-date?

Looking into removing mutual exclusion, and I have updated the description.

tldahlgren commented on lib/spack/spack/cmd/config.py:
```diff
@@ -68,6 +68,18 @@ def setup_parser(subparser):
+    list_scopes_parser = sp.add_parser("list-scopes", help="list defined scopes")
+    stype = list_scopes_parser.add_mutually_exclusive_group(required=False)
+    stype.add_argument(
+        "--file",
```
Please add support for `-f` as short-hand for this option.

greenc-FNAL committed:
`--file` and `--non-platform` are no longer mutually-exclusive

greenc-FNAL committed:
Add `-f` alias for `--file` option

greenc-FNAL committed:
Remove test dependence on external shell environment

greenc-FNAL committed:
Add tests for platform-specific scopes

greenc-FNAL committed:
Add test for `--file --non-platform`

tldahlgren commented on lib/spack/spack/test/cmd/config.py:
```diff
@@ -42,18 +43,35 @@ def test_config_list_scopes():
     output = config("list-scopes")
     assert "command_line" in output
     assert "_builtin" in output
+    assert f"site/{spack.platforms.host()}" in output
```
You probably don't want to add this. The session-level fixture `_use_test_platform()` causes the left hand side to be `site/test`.

greenc-FNAL replied:
What do you mean by, "session-level?" Is that unconditionally invoked for all tests, or as part of an in-test-file directive, or in test configuration, or on the command line? I'm not currently seeing the invocation mechanism, sorry.

tldahlgren replied:
They apply to all test functions run during the test "session" (https://docs.pytest.org/en/stable/explanation/fixtures.html).

tldahlgren commented on lib/spack/spack/test/cmd/config.py:
```diff
@@ -42,18 +43,35 @@ def test_config_list_scopes():
+
+
+def test_config_list_scopes_file_non_platform():
+    del os.environ["SPACK_DISABLE_LOCAL_CONFIG"]
```
This will fail on a key error.

greenc-FNAL replied:
Argh, thanks.

greenc-FNAL committed:
Update shell completion files

greenc-FNAL committed:
Fix key error in test

greenc-FNAL committed:
Use test platform in tests

greenc-FNAL committed:
Remove faulty list-scopes assertions

greenc-FNAL committed:
New `list-scopes` option `--included`

greenc-FNAL commented:
@tldahlgren I think I've addressed your comments, but if you have any tips on how to test either the inclusion/exclusion of platform scopes, or `--included`, I'd appreciate it.

greenc-FNAL committed:
Update command completion

greenc-FNAL committed:
Remove non-working test code

tldahlgren commented on lib/spack/spack/cmd/config.py:
```diff
@@ -68,6 +68,24 @@ def setup_parser(subparser):
 
     sp.add_parser("list", help="list configuration sections")
 
+    list_scopes_parser = sp.add_parser("list-scopes", help="list defined scopes")
```
Can  this be simplified to `scopes` so `spack config scopes`?

tldahlgren replied:
Thanks!

greenc-FNAL commented:
This will need to be adjusted when platform scopes go away with #50603.

greenc-FNAL committed:
`list-scopes` -> `scopes`

Per https://github.com/spack/spack/pull/41455#discussion_r2101321371.

greenc-FNAL committed:
Remove `--non-platform` in antipication of #50603

tldahlgren commented:
> Also @tldahlgren @tgamblin Any chance this can be in 1.0? I think this would complement #50603 nicely.

Yes. I didn't get a chance yesterday to look closely at the alternative for fixing the unit test. But I'm planning to do that today.

greenc-FNAL commented:
Would it be reasonable to take optional positional arguments which cause `spack config scopes` to list all the scopes that apply to a specific config type? Then `--file` would cause `include:almalinux9` to be displayed as `$spack/etc/spack/linux/almalinux9/packages.yaml` if that file were present when `spack config scopes --file packages` were invoked.

tgamblin commented:
could we make it `-p / --paths` to be consistent with `spack find --paths`?  And maybe when you supply the arg it displays the paths next to the name?

greenc-FNAL committed:
file -> path

greenc-FNAL committed:
Rationalize options and add ability to specify sections

* Improve help text for `spack config scopes`

* New option `--show-paths|-s` to show paths where scopes have them

* Add shortcut `-i` to `--include`

* Add non-option argument to specify section for which to show paths

* `--path` -> `--path-scopes` to avoid confusion with `--show-paths`

greenc-FNAL committed:
Remove unhelpful type hint

tldahlgren commented on lib/spack/spack/test/cmd/config.py:
```diff
@@ -40,6 +40,36 @@ def config_yaml_v015(mutable_config):
+def test_config_scopes_path_section():
+    output = config("scopes", "--include", "--path", "config")
+    assert "site" in output
+    assert "_builtin" not in output
```
It would be pretty straightforward to parameterize at least a couple of the new tests based on the (additional) options passed to the command, especially without additional checks.

On the chance you haven't done this before, there are two other tests in this module you can look to for reference (e.g., https://github.com/spack/spack/blob/c2e3b086fe2b22aab0d66d0c3cdc660de66761d6/lib/spack/spack/test/config.py#L480).

tldahlgren commented on lib/spack/spack/test/cmd/config.py:
```diff
@@ -40,6 +40,36 @@ def config_yaml_v015(mutable_config):
+
+def test_config_scopes_include():
+    output = config("scopes", "--include")
+    assert "site" in output
+    assert "_builtin" not in output
```
This would be more meaningful IMO if it checked to make sure that the scopes that are output all start with `include` (or `include:`).

tldahlgren replied:
Given `--include` takes precedence over `---show-paths`, the same goes for `test_config_scopes_include_path`.

greenc-FNAL replied:
Of course, as soon as #50603 merges, this will no longer be true :-(

greenc-FNAL replied:
The test now ensures that all included scopes contain `:`.

tldahlgren commented on lib/spack/spack/cmd/config.py:
```diff
@@ -215,6 +243,33 @@ def config_list(args):
+def _config_scope_info_string(args, scope):
+    if args.show_paths and hasattr(scope, "path"):
+        section_path = scope.get_section_filename(args.section) if args.section else None
+        path = section_path if section_path and os.path.exists(section_path) else f"{scope.path}/"
```
Should this be using `os.sep` in the `else` instead of `/`?

tldahlgren commented on lib/spack/spack/cmd/config.py:
```diff
@@ -215,6 +243,33 @@ def config_list(args):
+    """List configured scopes in descending order of precedence."""
+
+    if args.section and not args.show_paths:
+        tty.warning(f"[section] ({args.section}) ignored without --show-path")
```
Unfortunately, `tty` uses `warn` instead of `warning`:

```suggestion
        tty.warn(f"[section] ({args.section}) ignored without --show-paths")
```

tldahlgren replied:
Would also like to see the shortcut (`-s`) here.

tldahlgren replied:
Thanks

tldahlgren reviewed (changes requested):
See comments. Would like to see a unit test for the `-s | --show-paths` option. IMO it would be sufficient to make sure `os.path.join("etc", "spack", "defaults")` appears in the output.

greenc-FNAL committed:
Per @tldahlgren review

Fix style

tldahlgren commented on lib/spack/spack/cmd/config.py:
```diff
@@ -68,6 +68,34 @@ def setup_parser(subparser):
+    scopes_parser = sp.add_parser(
+        "scopes", help="list defined scopes in descending order of precedence"
+    )
+    scopes_parser.add_argument(
+        "-i", "--included", action="store_true", default=False, help="list only included scopes"
+    )
+    scopes_parser.add_argument(
+        "-p",
+        "--path-scopes",
+        action="store_true",
+        default=False,
+        help="list only writable scopes with an associated path",
+    )
```
Almost forgot to mention you can use the `add_mutually_exclusive_group()` method to allow the parser to handle the mutual exclusion for the options for you. Adding for these two arguments, you get:

```
$ spack config scopes -h
usage: spack config scopes [-hs] [-i | -p] [section]

positional arguments:
  section            tailor scope path information to the specified section
                     options: compilers, concretizer, definitions, env_vars, include, view, develop, mirrors, repos, packages, modules, config, upstreams, bootstrap, ci, cdash

options:
  -h, --help         show this help message and exit
  -i, --included     list only included scopes
  -p, --path-scopes  list only writable scopes with an associated path
  -s, --show-paths   show associated paths for appropriate scopes

$ spack config scopes -i -p
usage: spack config scopes [-hs] [-i | -p] [section]
spack config scopes: error: argument -p/--path-scopes: not allowed with argument -i/--included
```

greenc-FNAL replied:
I had mutually exclusive groups way back when, and then I removed them based on a suggestion from you (although I think the semantics of the relevant options were different). Included scopes are also path scopes, so when `-i` is specified the `-p` is redundant rather than inconsistent. I can certainly have them mutually exclusive in this case if you prefer, though.

What I'd like is for the non-option argument to be bound to the `-s` option if there's a way to do that, although having the presence of the non-option argument _imply_ `-s` would probably be sufficient.

greenc-FNAL committed:
Improve tests per @tldahlgren suggestions

greenc-FNAL committed:
Attempt to accommodate Windows™

greenc-FNAL committed:
Non-option argument `<section>` implies `-s|--show-paths`

greenc-FNAL commented:
@tldahlgren pending your re-review, and Gitlab getting around to running the CI pipeline, I think this is ready to go.

greenc-FNAL commented:
@spackbot run pipeline

tldahlgren reviewed (approved):
LGTM.
